doc/developer/design/20230418_subscribe_output.md

```diff
@@ -0,0 +1,282 @@
+- Feature name: Subscribe outputs
+- Associated: [#10593](https://github.com/MaterializeInc/materialize/issues/10593)
+
+# Summary
+[summary]: #summary
+
+Clients have had trouble using subscribe. They've struggled
+reconciling the current output of subscribe, a list of inserts and
+deletes of tuples, with their application-level data. To
+bridge that gap we introduce three ways to change how subscribe
+produces its output: `WITHIN TIMESTAMP ORDER BY`, `ENVELOPE UPSERT`,
+and `ENVELOPE DEBEZIUM`. These output types are simple to implement
+because they operate on the diffs present in a single timestamp. We
+hope they'll match the client's mental model of updates on their results.
+
+# Motivation
+[motivation]: #motivation
+
+Sometimes clients want to think of the output of subscribe as a list of
+updates to the underlying relation. While the current output reflects
+the changes to the underlying relation, the lack of consistent
+ordering has troubled some clients. They worry about handling a case
+like `(k, v2, +1), (k, v1, -1)` where an update appears as an insert
+for the new value for a particular key occurs followed by the deletion
+of its previous value. The output types proposed all solve this
+problem with varying degrees of flexibility. `ENVELOPE UPSERT` will
+transform the output of `SUBSCRIBE` into a series of `UPSERTS`;
+`ENVELOPE DEBEZIUM` will transform the output into a series of
+debezium updates with a `before` and `after` state. When viewed as
+upserts the ordering of retractions and insertions doesn't
+matter. `WITHIN TIMESTAMP ORDER BY` is the most flexible: the client
+can specify a set of columns to order the output by (within each
+timestamp), including `mz_diff`. They can use this ordering to force
+retractions to occur before additions in order to simplify the client
+side handling.
+
+For the simple case of a one to one mapping between keys and values
+(there's at most one value at a time for each key), any of these
+output types would let the client see the changes to the underlying
+relation as a series of upserts. `WITHIN TIMESTAMP ORDER BY` could
+also be used to sort the results from `SUBSCRIBE`; we can think of the
+results from `SUBSCRIBE` are naturally ordered by `mz_timestamp`, we
+could supply a secondary order. I don't think there are many real use
+cases that would benefit from this.
+
+
+# Explanation
+[explanation]: #explanation
+
+`WITHIN TIMESTAMP ORDER BY` takes a `ORDER BY` expression and sorts
+the returned data within each distinct timestamp. This `ORDER BY` can
+take any column in the underlying object or query in addition to the
+`mz_diff` column. A common use case of recovering upserts to key-value
+data would be sort by all the keys and `mz_diff`. Since `SUBSCRIBE`
```

Comment:
One idea, and maybe a bad idea, is that this should be the default sort order.

Reply:
I think we don't know what the keys are for the underlying object unless specified, so aren't able to do that currently, but I could be wrong.

Reply:
yeah, this notion of key only exists at the subscribe statement